### ISSUE TITLE
Allow react-native & expo to be part of devDependencies

### ISSUE THREAD

**Issue by fabiancook** [expo/expo-cli#96]
## Description

Allow `react-native` & `expo` to be part of `devDependencies` instead of only `dependencies`, this would allow developers to create & test interfaces using expo, export the project on npm, and then include it in another project where native dependencies can be utilised. 

If `react-native` is in the `devDependencies` then we will be able to utilise a different version of `react-native`, given we are using different version compatible APIs. 

## Expected Behavior

`expo publish` to work with no errors given `react-native` & `expo` are in `devDependencies`, and have been installed in my local `node_modules`

## Observed Behavior

`expo publish` informs me that `react-native` is not in my `dependencies` 

```
$ expo publish
[23:25:55] Unable to find an existing Expo CLI instance for this directory, starting a new one...
[23:25:57] Error: Can't find react-native in package.json dependencies
[23:25:59] Starting Metro Bundler on port 19001.
[23:26:00] Metro Bundler ready.
[23:26:02] Tunnel ready.
[23:26:02] Publishing to channel 'default'...
[23:26:04] Error: Can't find react-native in package.json dependencies
[23:26:04] Couldn't publish because errors were found. (See logs above.) Please fix the errors and try again.
```

## Environment

```$ expo diagnostics
[23:35:13] Generating diagnostics report...
[23:35:13] You can join our slack here: https://slack.expo.io/.

Environment:
  OS: macOS High Sierra 10.13.4
  Node: 10.10.0
  Yarn: 1.9.4
  npm: 6.4.1
  Watchman: 4.9.0
  Xcode: Xcode 9.4.1 Build version 9F2000
  Android Studio: 3.0 AI-171.4443003

Packages: (wanted => installed)
  expo: ^30.0.1 => 30.0.1
  react: 16.3.1 => 16.3.1
  react-native: https://github.com/expo/react-native/archive/sdk-30.0.0.tar.gz => 0.55.4

Diagnostics report:
  https://exp-xde-diagnostics.s3.amazonaws.com/harvest-snapp-d4872c8d-24a2-4b77-bae5-8a08d078bc13.tar.gz
```

**Comment by alaboudi** [comment]
I think you have an incorrect use case for devDependencies. devDependency packages are not meant to be part of a build. I think the way they have it currently makes most semantic sense.

**Comment by fabiancook** [comment]
Correct, devDependencies are not meant to be part of the build, I'm using expo for development & "building" it as a standard npm module which can then be used by another react native project. With this in mind I don't want expo or react native to be part of my main dependencies when I publish the module to npm.

**Comment by brentvatne** [comment]
@fabiancook - can you share the project? I'm not sure I understand the setup you describe

**Comment by dagatsoin** [comment]
@brentvatne @alaboudi I face a similar problem. I want to exclude React Native and Expo from the published npm package. Here is my working repo https://github.com/dagatsoin/sproutch

If I move react-native and expo in devDeps I face the same error as @fabiancook

**Comment by brentvatne** [comment]
you probably need to change the project structure to separate the storybook/example content from the library itself, eg: https://github.com/react-navigation/react-navigation-stack has an example directory

**Comment by fabiancook** [comment]
I think the thing we want to know is why `react-native` and `expo` **need** to be `dependencies`, and why it can't be either `dependencies` or `devDependencies`? 

In our cases both `react-native` and `expo` are only needed in our development environment or build systems, they are not required when we publish our module, as the correct version of `react-native` and `expo` are being maintained somewhere else. 

In my case I don't use `expo` at all in the production app, `expo` is only being used to develop the views being used by the application. 

--------------------------------

TLDR: remove the hard requirement for `react-native` and `expo` to be in `dependencies` and allow it to be in either `dependencies` or `devDependencies`

**Comment by brentvatne** [comment]
cc @fson seems reasonable to support this usecase

**Comment by fabiancook** [comment]
Would I be correct in saying this is the only points where this is checked...

https://github.com/expo/expo-cli/blob/140a27f37d7e7ff353eb6a043d8fad811df521c9/packages/xdl/src/project/Doctor.js#L303-L315

https://github.com/expo/expo-cli/blob/140a27f37d7e7ff353eb6a043d8fad811df521c9/packages/xdl/src/project/Doctor.js#L474-L481

**Comment by fabiancook** [comment]
`peerDependencies` also seems like a good candidate to be supported as suggested by another user (They had deleted their comment on the PR)

**Comment by brentvatne** [comment]
closing this since the pr is open and all we need to do is resolve conflicts and we can merge :)

**Comment by tpmai22** [comment]
When install `expo-cli` as a `devDependecies` would it have any draw back that I should be aware of ?
